### PR TITLE
[IMM32] Improve ImmGetImeInfoEx

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Implementing Japanese input...
JIRA issue: [CORE-11700](https://jira.reactos.org/browse/CORE-11700)

## Proposed changes

- Rewrite `ImmGetImeInfoEx` function.
- Implement `CtfImmIsTextFrameServiceDisabled` function.
- Modify `imm32.spec`.

### CODE REVIEW

File: dll/win32/imm32/imm.c [hunk 1]
```diff
@@ -3608,29 +3608,67 @@ BOOL WINAPI ImmRegisterClient(PVOID ptr, /* FIXME: should point to SHAREDINFO st
     return TRUE;
 }
 
+/***********************************************************************
+ *		CtfImmIsTextFrameServiceDisabled(IMM32.@)
+ */
+BOOL WINAPI CtfImmIsTextFrameServiceDisabled(VOID)
+{
+    PTEB pTeb = NtCurrentTeb();
+    if (pTeb->Win32ClientInfo[0] & 0x400)
```
Comment (HBelusca) [inline dll/win32/imm32/imm.c:3617]: A flag name will certainly be needed there.

Comment (HBelusca) [reply dll/win32/imm32/imm.c:3617]: Also please use something like:
`((PW32CLIENTINFO)NtCurrentTeb()->Win32ClientInfo)->WhateverFieldIsThere`
(you need to check which header(s) you need to include for getting `PW32CLIENTINFO`; I know gdi32 includes it)

And, just do: `return (((PW32CLIENTINFO)NtCurrentTeb()->Win32ClientInfo)->WhateverFieldIsThere & 0x400);`

File: dll/win32/imm32/imm.c [hunk 2]
```diff
@@ -3608,29 +3608,67 @@ BOOL WINAPI ImmRegisterClient(PVOID ptr, /* FIXME: should point to SHAREDINFO st
...
+
+    hKL = *(HKL*)pvSearchKey;
+    pImeInfoEx->hkl = hKL;
+
+    if (!IS_IME_HKL(hKL))
+    {
+        if (g_dwImm32Flags & IMM32_FLAG_CICERO_ENABLED)
+        {
+            pTeb = NtCurrentTeb();
+            if (pTeb->Win32ClientInfo[5] & 2)
```
Comment (HBelusca) [inline dll/win32/imm32/imm.c:3664]: See comment above.

**Review by HBelusca — CHANGES_REQUESTED** [review]

File: dll/win32/imm32/imm.c (hunk 1 above)
Comment (katahiromz) [reply dll/win32/imm32/imm.c]: OK.

File: dll/win32/imm32/imm.c (hunk 2 above)
Comment (katahiromz) [reply dll/win32/imm32/imm.c]: This is currently unknown entirely.

**Pull request merged**